File: python/day2.py

inp = []

with open("../input/input02.txt") as f:
    inp = [line.strip() for line in f]

def tally(s):
    counts = {}
    for c in s:
        counts[c] = counts.get(c, 0) + 1
    return counts

def commonChars(s1, s2):
    chars = [c1 for (c1, c2) in zip(s1, s2) if c1 == c2]
    return "".join(chars)

def part1(ids):
    count2 = 0
    count3 = 0

    for id_ in ids:
        counts = tally(id_)
        if any([n == 2 for n in counts.values()]):
            count2 += 1
        if any([n == 3 for n in counts.values()]):
            count3 += 1

    return count2 * count3
        
def part2(ids):
    for id1 in ids:
        for id2 in ids:
            common = commonChars(id1, id2) 
            if len(common) == len(id1) - 1:
                return common

print part1(inp)
print part2(inp)
            

File: haskell/Day2.hs
module Main (main) where

import AOC.Input
import AOC.Misc (count, tally)

year = 2018
day  = 2

parseInput :: String -> [String] 
parseInput = lines

part1 :: [String] -> Int
part1 ss = count2 * count3
  where
    counts = map tally ss
    count2 = count (any (\(_, n) -> n == 2)) counts
    count3 = count (any (\(_, n) -> n == 3)) counts

part2 :: [String] -> String
part2 ss = head [common | s1 <- ss,
                          s2 <- ss,
                          let common = [c1 | (c1, c2) <- zip s1 s2, c1 == c2],
                          length common == length s1 - 1]

main :: IO ()
main = do
  input <- parseInput <$> getInput year day
  print $ part1 input
  print $ part2 input
